### PR TITLE
Copter: Add new TURTLE mode - Flip over after crash

### PR DESCRIPTION
**Pull request by andyp1per**
Adds a new mode which allows the user to reverse and spin up two motors in order to flip over a copter after a crash. Which two motors are spun up is controlled by roll and pitch and the greater the roll and pitch input the higher the throttle. This mode only works when the copter is disarmed as it provides no control. Needs to be used with care since a mode that allows motor spin up while disarmed has some potential safety issues. Centered sticks stops all motors.

This is implemented as a flight mode and also RC function 107.

Demo:
https://youtu.be/LdQW4aBnItg

### CODE REVIEW

**Comment by andyp1per** [comment]
Fixes https://github.com/ArduPilot/ardupilot/issues/12109

File: ArduCopter/mode_turtle.cpp [hunk 1]
```diff
@@ -0,0 +1,138 @@
+#include "Copter.h"
+
+#if MODE_TURTLE_ENABLED == ENABLED
+
+#define CRASH_FLIP_EXPO 35.0f
+#define CRASH_FLIP_STICK_MINF 0.15f
+#define power3(x) ((x)*(x)*(x))
+
+// throw_init - initialise turtle mode controller
```
Comment (Gone4Dirt) [inline ArduCopter/mode_turtle.cpp:9]: is this a copy-paste mistake? What do you mean by throw_init?

File: ArduCopter/mode_turtle.cpp [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    // do not enter the mode when already armed or when flying
+    if (motors->armed() || SRV_Channels::get_dshot_esc_type() == 0) {
+        return false;
+    }
+    // reverse the motors
+    hal.rcout->disable_channel_mask_updates();
+    change_motor_direction(true);
+
+    // arm
+    motors->armed(true);
```
Comment (Gone4Dirt) [inline ArduCopter/mode_turtle.cpp:21]: Do we really want to arm the motors just by changing into the flight mode?  I think we should force users to arm again if disarmed.

File: ArduCopter/mode_turtle.cpp [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+    return true;
+}
+
+void ModeTurtle::exit()
+{
+    // disarm
+    motors->armed(false);
+    hal.util->set_soft_armed(false);
+
```
Comment (Gone4Dirt) [inline ArduCopter/mode_turtle.cpp:32]: Same as above. Arming and disarming from mode entry and exit is not a good idea IMHO

File: ArduCopter/mode_turtle.cpp [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+#define CRASH_FLIP_EXPO 35.0f
+#define CRASH_FLIP_STICK_MINF 0.15f
+#define power3(x) ((x)*(x)*(x))
+
+// throw_init - initialise turtle mode controller
+bool ModeTurtle::init(bool ignore_checks)
+{
+    // do not enter the mode when already armed or when flying
+    if (motors->armed() || SRV_Channels::get_dshot_esc_type() == 0) {
```
Comment (tridge) [inline ArduCopter/mode_turtle.cpp:13]: check sticks centered? within DZ?

**Review by tridge — CHANGES_REQUESTED** [review]

File: ArduCopter/mode_turtle.cpp (hunk 1 above)
Comment (andyp1per) [reply ArduCopter/mode_turtle.cpp:9]: Yes copy and paste!

Comment (andyp1per) [reply ArduCopter/mode_turtle.cpp]: Fixed

File: ArduCopter/mode_turtle.cpp (hunk 2 above)
Comment (andyp1per) [reply ArduCopter/mode_turtle.cpp:31]: Arming checks added

File: ArduCopter/mode_turtle.cpp (hunk 3 above)
Comment (andyp1per) [reply ArduCopter/mode_turtle.cpp:47]: Arming checks added

File: ArduCopter/mode_turtle.cpp (hunk 4 above)
Comment (andyp1per) [reply ArduCopter/mode_turtle.cpp:12]: Fixed

**Review by tridge — DISMISSED** [review]
we decided the arming checks are not needed for turtle mode

**Comment by rmackay9** [comment]
I'm happy for this to go in after the above changes are made so let's not hold this up too much longer.  @peterbarker, etc feel free to merge after @andyp1per's update if I don't see it.

File: ArduCopter/mode_turtle.cpp [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+                hal.rcout->send_dshot_command(inverse_direction, i, 0, 10, true);
+            }
+        }
+    }
+}
+
+void ModeTurtle::run()
+{
+    const float flipPowerFactor = 1.0f - CRASH_FLIP_EXPO / 100.0f;
+    const float stickDeflectionPitch =  channel_pitch->norm_input_dz();
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply ArduCopter/mode_turtle.cpp]: Fixed

File: libraries/RC_Channel/RC_Channel.h [hunk 6]
```diff
@@ -208,6 +208,7 @@ class RC_Channel {
         EKF_YAW_RESET =      104, // trigger yaw reset attempt
         GPS_DISABLE_YAW =    105, // disable GPS yaw for testing
         DISABLE_AIRSPEED_USE = 106, // equivalent to AIRSPEED_USE 0
+        TURTLE =             107, // Turtle mode - flip over after crash
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/RC_Channel/RC_Channel.h:211]: this is probably a good idea.  We seem to be reserving the 10x values for disabling things..

[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/RC_Channel/RC_Channel.h]: ok Fixed

**Pull request merged**